plugins/flytekit-comet-ml/flytekitplugins/comet_ml/tracking.py

```diff
@@ -0,0 +1,146 @@
+
+import comet_ml
+from flytekit import Secret
+from flytekit.core.context_manager import FlyteContextManager
+from flytekit.core.utils import ClassDecorator
+
+COMET_ML_EXECUTION_TYPE_VALUE = "comet-ml-execution-id"
+COMET_ML_CUSTOM_TYPE_VALUE = "comet-ml-custom-id"
+
+
+def _generate_suffix_with_length_10(project_name: str, workspace: str) -> str:
+    """Generate suffix from project_name + workspace."""
+    h = shake_256(f"{project_name}-{workspace}".encode("utf-8"))
+    # Using 5 generates a suffix with length 10
+    return h.hexdigest(5)
+
+
+def _generate_experiment_key(hostname: str, project_name: str, workspace: str) -> str:
+    """Generate experiment key that comet_ml can use:
+
+    1. Is alphanumeric
+    2. 32 <= len(experiment_key) <= 50
+    """
+    # In Flyte, then hostname is set to {.executionName}-{.nodeID}-{.taskRetryAttempt}, where
+    # - len(executionName) == 20
+    # - 2 <= len(nodeId) <= 8
+    # - 1 <= len(taskRetryAttempt)) <= 2 (In practice, retries does not go above 99)
+    # Removing the `-` because it is not alphanumeric, the 23 <= len(hostname) <= 30
+    # On the low end we need to add 10 characters to stay in the range acceptable to comet_ml
+    hostname = hostname.replace("-", "")
+    suffix = _generate_suffix_with_length_10(project_name, workspace)
+    return f"{hostname}{suffix}"
+
+
+class comet_ml_init(ClassDecorator):
+    COMET_ML_PROJECT_NAME_KEY = "project_name"
+    COMET_ML_WORKSPACE_KEY = "workspace"
+    COMET_ML_EXPERIMENT_KEY_KEY = "experiment_key"
+    COMET_ML_URL_SUFFIX_KEY = "link_suffix"
+    COMET_ML_HOST_KEY = "host"
+
+    def __init__(
+        self,
+        task_function: Optional[Callable] = None,
+        project_name: Optional[str] = None,
+        workspace: Optional[str] = None,
+        experiment_key: Optional[str] = None,
+        secret: Optional[Union[Secret, Callable]] = None,
+        host: str = "https://www.comet.com",
+        **init_kwargs: dict,
+    ):
+        """Comet plugin.
+        Args:
+            task_function (function, optional): The user function to be decorated. Defaults to None.
+            project_name (str): Send your experiment to a specific project. (Required)
+            workspace (str): Attach an experiment to a project that belongs to this workspace. (Required)
+            experiment_key (str): Experiment key.
+            secret (Secret or Callable): Secret with your `COMET_API_KEY` or a callable that returns the API key.
+                The callable takes no arguments and returns a string. (Required)
+            host (str): URL to your Comet service. Defaults to "https://www.comet.com"
+            **init_kwargs (dict): The rest of the arguments are passed directly to `comet_ml.init`.
+        """
+        if project_name is None:
+            raise ValueError("project_name must be set")
+        if workspace is None:
+            raise ValueError("workspace must be set")
+        if secret is None:
```

Comment:
shouldn't `secret` be made optional as it isn't necessary when executing locally?

Reply:
It's ignored during location execution, but required during remote. If this was optional, we'll catch the "no secret" during execution and not at registration time. I do not want to wait until then, because these experiments can use GPUs and waiting & spinning up a GPU just to see this error is not efficient and a worst DevX.

By making `secret` required, I am optimizing for "failing quickly" for remote execution.